### PR TITLE
Datetime column, removing short trips, and historical data scripts

### PR DESCRIPTION
**Pull request by Charlie-Henry**
https://github.com/cityofaustin/atd-data-tech/issues/12705

Improvements to the [dataset](https://data.austintexas.gov/Transportation-and-Mobility/Austin-MetroBike-Trips/tyfh-5r8s/data):
- Reprocessed all historical data to populate `bike_type` column. So now all trips are either `electric` or `classic` (human) powered trips.
- Added a `checkout_datetime` column that combines the `checkout_date` and `checkout_time` (those remain in the dataset for backwards compatibility) columns so it is much easier to query dates and times together. `checkout_time` was just a text field so it was difficult to query with that.
- I've deleted all trips under 2 minutes in length. Hunter (CapMetro) has been removing these trips himself recently but I noticed we had a bunch in the historical data that I've removed, and I've enforced that we ignore those records with a trip duration of less than 2 minutes going forward.

This PR also contains two scripts I used to re-process the archive of MetroBike data. 
- `archive_processing_2013-2015.py` handles data that was provided to us by MetroBike corporate that has some significant differences with the regular export from MetroBike we have at other time periods. This new file had to be created in our previous archive of data did not contain a `trip_id` column so it much harder to update things and remove errors in our data. 
- `archive_processing_2016-2022.py` is similar to our current production ETL but ignores the field schema validation John implemented to the main ETL since these files have nulls where it was previously acceptable to have null fields.
- I do use pandas in these ☝️ but I don't believe we need to include that in `requirements.txt` since these files should be run locally anyway and not needed to be included with any of our MetroBike data pipelines. Maybe a separate `requirements.txt` just for that directory?

Some things to address in the future:
- Cleaning up the kiosks [dataset](https://data.austintexas.gov/Transportation-and-Mobility/Austin-MetroBike-Kiosk-Locations/qd73-bsdg/data) and validate the kiosk IDs in our trips dataset has a matching kiosk.
- I noticed in the corporate MetroBike dataset, that there is a trip distance field. This is GPS distance the trip took so that may be interesting to include.

### CODE REVIEW

File: publish_trips.py [hunk 1]
```diff
@@ -161,6 +174,10 @@ def handle_data(csv_text):
     data = classify_bike_type(data)
     # Add month/year columns
     data = populate_month_year(data)
+    # Add datetime column
+    data = datetime_creation(data)
+    # Remove trips that are less than 2 minutes in length
+    data = [d for d in data if int(d.get("trip_duration_minutes", 0)) > 1]
```
Comment (Charlie-Henry) [inline publish_trips.py:177]: Hunter has been adding this filter manually to the data he provides to us but it is good practice to make sure we drop them to keep things consistent.

File: historical_data_processing/archive_processing_2013-2015.py [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+kiosks.loc[kiosks["kiosk_id"] == 2546, "city_asset_num"] = 2546
+kiosks.loc[kiosks["kiosk_id"] == 2545, "city_asset_num"] = 2545
+kiosks.loc[kiosks["kiosk_id"] == 2712, "city_asset_num"] = 6
+kiosks.loc[kiosks["kiosk_id"] == 1006, "city_asset_num"] = 1006
+kiosks.loc[kiosks["kiosk_id"] == 2500, "city_asset_num"] = 2500
+
+# Create lookup table between city_asset_num used by the CSV and kiosk_ids we use in the dataset
+kiosks = kiosks[kiosks["city_asset_num"] >= 0]
+kiosks["city_asset_num"] = kiosks["city_asset_num"].astype("Int32")
+kiosks = kiosks[["kiosk_id", "city_asset_num"]]
```
Comment (frankhereford) [inline historical_data_processing/archive_processing_2013-2015.py:115]: 🐼 💪 - how this is rolling up these loops to slice and dice the data is really concise and readable. Really nice - TIL.

**Review by frankhereford — APPROVED** [review]
I have not run these locally, but I have read the code, and it all looks good. Especially with the archive processing script, you're doing a lot of the T in ETL!

What do you think about incorporating the 1Password API usage pattern into these?

File: historical_data_processing/archive_processing_2013-2015.py [hunk 3]
```diff
@@ -0,0 +1,229 @@
+"""
+Script that publishes data to Socrata from a file that is stored on Dropbox that contains data for
+trips between December 2013 and January 2016. This file is in a different format with different column
+definitions so that's why this script needed to be created.
```
Comment (johnclary) [inline historical_data_processing/archive_processing_2013-2015.py:4]: 👍

File: publish_trips.py [hunk 4]
```diff
@@ -151,6 +152,18 @@ def populate_month_year(data):
...
+    """
+    Joins together the checkout time and date columns to a single datetime field.
+    """
+    for row in data:
+        # Hour is not zero padded in the data so we have to check for that
+        if len(row["checkout_time"]) == 7:
+            row["checkout_datetime"] = f"{row['checkout_date']}T0{row['checkout_time']}"
+        else:
+            row["checkout_datetime"] = f"{row['checkout_date']}T{row['checkout_time']}"
+    return data
```
Comment (johnclary) [inline publish_trips.py:162]: just thinking about data validation, if this datetime string ends up being an invalidate date, Socrata will hopefully throw an error, right?

what you have is great but `zfill` is a fun string method:
```python
>>> "1:30:00".zfill(8)
'01:30:00'
```

File: publish_trips.py [hunk 5]
```diff
@@ -161,6 +174,10 @@ def handle_data(csv_text):
     data = classify_bike_type(data)
     # Add month/year columns
     data = populate_month_year(data)
+    # Add datetime column
+    data = datetime_creation(data)
```
Comment (johnclary) [inline publish_trips.py]: i know i established this pattern, but here's something i've been trying to get more consistent about: since each of the three functions above modify the rows in-place, it'd probably be better form to not reassign `data` after each call. so this instead:


```python
classify_bike_type(data)
# Add month/year columns
populate_month_year(data)
# Add datetime column
datetime_creation(data)
```

**Review by johnclary — APPROVED** [review]
@Charlie-Henry it's so awesome that you have taken on this data cleanup! this was a lot of work and the dataset is so much better for it. 

i have merely picked nits in my review—this is good to go! 🚢 

wrt to filtering out short trips, i'd prefer that we share the unfiltered data. that's what we did with the [scooter trip data](https://data.austintexas.gov/Transportation-and-Mobility/Shared-Micromobility-Vehicle-Trips/7d8e-dm7r), and we added a note explaining that our official trip reporting excluded trips based on time + distance.

however, since Hunter is doing the filtering upstream i agree it's best if we're consistent. would you mind adding a note to the dataset description that mentions this filter?

wrt to incorporating 1password per @frankhereford's suggestion: i'm of two minds about it. my instinct is to leave the 1pass integration to our orchestration tool and continue to keep our code agnostic. i guess it just "feels" like an orchestration-level concern to me? i would not want to be the person trying to track down all of our 1pass dependencies if we need to move away from it at some point. 

that said, i think i'm really exaggerating the overhead of adding 1pass directly into our scripts. if we're going to keep going that route, we might should publish our Airflow 1pass utility to pypi? otherwise we're repeating a lot of boilerplate. i dunno!

🚢 🚢 🚢 🚢 🚢 🚢

File: publish_trips.py (hunk 4 above)
Comment (Charlie-Henry) [reply publish_trips.py:162]: Ah TIL about `zfill`!

Socrata will not error out based on my experience if the datetime is an invalid date. I think it'll usually just end up being nulled 👎

Comment (Charlie-Henry) [reply publish_trips.py:162]: Now that I think about about some more, we could potentially try to use the Socrata [dataset alerts](https://support.socrata.com/hc/en-us/articles/360000235507-Create-and-Manage-Dataset-Alerts-on-the-Enterprise-Data-Platform) to try to catch these bugs as they enter the dataset in addition to cerberus in the ETL. Should hopefully let us stay on top of these rather than letting them build dust potentially?

**Comment by frankhereford** [comment]
@johnclary wrote:

> wrt to incorporating 1password per @frankhereford's suggestion: i'm of two minds about it. my instinct is to leave the 1pass integration to our orchestration tool and continue to keep our code agnostic. i guess it just "feels" like an orchestration-level concern to me? i would not want to be the person trying to track down all of our 1pass dependencies if we need to move away from it at some point.

> that said, i think i'm really exaggerating the overhead of adding 1pass directly into our scripts. if we're going to keep going that route, we might should publish our Airflow 1pass utility to pypi? otherwise we're repeating a lot of boilerplate. i dunno!

You have good instincts about big picture stuff John. 

I can definitely see that the scripts become less portable / usable outside of DTS if we start pulling from 1PW directly instead of from the environment. This is ever so slightly contrary to open-by-default. 

But doing it makes it real easy to run them on the console for us, especially if we settle on a common set of environment variables that we use to store the 1PW API server and its related tokens, so we can put them in our `.zsh` configurations on our machines. 

There are lots of really valid opinions on both sides of the fence here, imho. Thank you @johnclary for your comments - you always get me thinking and that is awesome!

### TLDR

I'm not interested in pushing others on our team to adopt using it over tried and true techniques, at least not at this time. I think if it's a better way to do things, it will be better enough for us to start using it organically. I've had a good experience doing the VZ ETL using the 1PW library directly, and I'm excited to keep exploring it and finding new ways to leverage that 1PW server. I'm fully open to changing this and to discussing it more too if anyone wants to do that.

File: publish_trips.py (hunk 5 above)
Comment (frankhereford) [reply publish_trips.py]: Thanks for reminding me about this John. I forget that python will pass in a reference to mutable objects. Doing like you're saying would be a good habit for me to get into too.

**Review by mddilley — APPROVED** [review]
I read through this code, and it looks great! I did not attempt to run it in any way, but I really like how these scripts tell the story of how the data changed over time. 🚢 🚀

I really enjoyed the 1PW integration discussion in here, and I'd love to talk about in our Airflow check-in so I added it to the [ETL discussion doc](https://docs.google.com/document/d/1aoZJacu9XxmyOTKLY88l1SfjOeqiAzeJGXS96m5Snu8/edit). 🔐

File: historical_data_processing/archive_processing_2013-2015.py [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+df["RentalDate CST"] = pd.to_datetime(df["RentalDate CST"].str[:-7])
+df["RentalDate CST"] = (
+    df["RentalDate CST"].dt.tz_localize("utc").dt.tz_convert(central_timezone)
+)
+df["checkout_date"] = df["RentalDate CST"].dt.strftime("%Y-%m-%d")
+df["checkout_time"] = df["RentalDate CST"].dt.strftime("%-H:%M:%S")
+df["checkout_datetime"] = df["RentalDate CST"].dt.strftime("%Y-%m-%dT%H:%M:%S")
+df["year"] = df["RentalDate CST"].dt.strftime("%Y")
+df["month"] = df["RentalDate CST"].dt.strftime("%-m")
+df["hour"] = df["RentalDate CST"].dt.strftime("%H")
```
Comment (roseeichelmann) [inline historical_data_processing/archive_processing_2013-2015.py:95]: noiceee this seems annoying to have to convert from string to datetime back to string but u handled it well

File: publish_trips.py (hunk 4 above)
Comment (roseeichelmann) [reply publish_trips.py:162]: So would u use the socrata dataset alert to alert us when the datetime column is `null`?

**Review by roseeichelmann — APPROVED** [review]
Didn't test but the code looks good! This was a lot of work, thank you for keeping our datasets clean and up to date. You make it look easy!

File: publish_trips.py (hunk 4 above)
Comment (Charlie-Henry) [reply publish_trips.py:162]: @roseeichelmann yeah you can query the dataset and if it returns rows it will send an alert email. It's a pretty neat feature but I don't think it's super necessary here, sometime later prehaps.

**Comment by Charlie-Henry** [comment]
Updated the dataset description, thanks for the reviews yall.
> Note that trips less than 2 minutes are not included in this dataset. Rebalancing and maintenance trips by MetroBike staff are also not included.
